### ISSUE TITLE
Decorated property not supported

### ISSUE THREAD

**Issue by pyrocat101** [python/mypy#1362]
``` python
class Foo(object):
    @property
    @some_decorator
    def foo(self):
        ...
```

is valid in Python 2, but MyPy complains `Decorated property not supported`.
Adding `type: ignore` does not ignore the error.

**Comment by gvanrossum** [comment]
Good news: if you put the `# type: ignore` on the `@property` line it's silent.

**Comment by nullp0tr** [comment]
A bit off topic, but I came across this warning today and it made me wonder about something. I have a BLE Connection class, and each instance has the attributes characteristics and services, which are implemented with the @property decorator so that it can actually fetch them a new every time without assigning a lambda or making them a method. Fetching can raise an exception though, and I'm wondering how normal is it that something like "a = b.attribute" can raise an exception that can't be just avoided by rearranging the code like from a NameError or a SyntaxError?

**Comment by gvanrossum** [comment]
Since properties are common in Python, that's not something to be concerned about.

**Comment by nullp0tr** [comment]
Thanks for the clarification.

**Comment by OJFord** [comment]
> Good news: if you put the `# type: ignore` on the `@property` line it's silent.

Note it appears to need to be on the outer decorator, so if you have the opposite order to OP:
```
@some_decorator
@property
def foo
```
then it should be on the `@some_decorator` line.

**Comment by gvanrossum** [comment]
Correct, messages related to a function (or, sometimes, its arguments)
always appear on the line of the first decorator, if there is one,
otherwise on the linke containing 'def'. (I consider this a slight flaw, as
occasionally it makes it hard to find the root cause for an error message.)

The rule "put the '# type: ignore' on the line where the error appears" is
pretty simple to remember though.

**Comment by ZahlGraf** [comment]
Hello

I have the same issue. In my project I use additional decorators for properties very often. Thus mypy sends me hundreds of error with "error: Decorated property not supported".

First of all, I don't understand the error message. Does it mean mypy does not support decorated properties, or is decorating a property related to any code quality issue in general? Because if just mypy is not supporting decorated properties, but the pattern in general is not considered as bad practice, I don't know, why mypy is giving me an error here. It should be at maximum a warning right?

Second, if it is just a mypy issue, I really want to ignore it. Is there any way to ignore exactly this error on a global level? At the moment, I'm only aware of ignoring type errors for a single line, but since this pattern is extremely common in my code, I don't want to add the "ignore line" comment on hundreds of lines in my code.

Thanks in advance for any suggestions and assistance.

**Comment by JukkaL** [comment]
Mypy cannot properly type check code involving decorated properties. Currently there's no way to ignore all such errors, but this is something mypy may support in the future. 

As a workaround, you could try defining your own alias to the `property` decorator that is annotated to return `Any`. If you use your custom property decorator (that at runtime behaves exactly like `property`) for all decorated properties, mypy will silently give all such properties the `Any` type.

**Comment by kawing-chiu** [comment]
Same issue here. I have to use decorated property because the `abc` module have deprecated `@abstractproperty`. The only way to define an abstract property is to use decorated property:
```
class C(ABC):
    @property
    @abstractmethod
    def my_abstract_property(self):
        ...
    @my_abstract_property.setter
    @abstractmethod
    def my_abstract_property(self, val):
        ...
```
If generally decorated properties cannot be supported, can mypy special-case and support `@abstractmethod`?

**Comment by ilevkivskyi** [comment]
> If generally decorated properties cannot be supported, can mypy special-case and support `@abstractmethod`?

Yes, this (supporting abstract settable properties) is actually what is done by PR https://github.com/python/mypy/pull/8111 (it is a few line fix I added because this broke couple tests for the main PR matter).

**Comment by warsaw** [comment]
> Same issue here. I have to use decorated property because the `abc` module have deprecated `@abstractproperty`. The only way to define an abstract property is to use decorated property:
> 
> ```
> class C(ABC):
>     @property
>     @abstractmethod
>     def my_abstract_property(self):
>         ...
> 
>     @my_abstract_property.setter
>     @abstractmethod
>     def my_abstract_property(self, val):
>         ...
> ```
> 
> If generally decorated properties cannot be supported, can mypy special-case and support `@abstractmethod`?

This is exactly my use case too.

**Comment by analog-cbarber** [comment]
Wish there was some easy way to mark the decorator function to tell mypy to ignore it or a way to tell mypy to always ignore such decorators.

It is really ugly to have to slap a `# type: ignore` comment on every use of a property decorator.

Let's put it this way: what is the concrete benefit of the existing behavior?

**Comment by Fokko** [comment]
Same here :)

**Comment by MicaelJarniac** [comment]
What's the progress on this issue?

I'm using the `@functools.lru_cache` decorator on many `@property` methods and this "warning" shows up on them all.

```python
from functools import lru_cache

class MyClass:
    def __init__(self, year: int):
        self.current_year = 2020
        self.year = year

    @property  # Decorated property not supported - mypy(error)
    @lru_cache
    def how_long_ago(self) -> int:
        return self.current_year - self.year
```

I'm using the "workaround" of adding `# type: ignore`, but it's far from ideal.

```python
from functools import lru_cache

class MyClass:
    def __init__(self, year: int):
        self.current_year = 2020
        self.year = year

    @property  # type: ignore
    @lru_cache
    def how_long_ago(self) -> int:
        return self.current_year - self.year
```

`functools` should have a [`@cached_property`](https://docs.python.org/dev/library/functools.html?highlight=s#functools.cached_property) decorator that would probably not trigger this "warning" from `mypy`, but it's only for Python ≥ `3.8`, and I'm currently stuck on Python ≤ `3.7`, so I can't use it.

```python
from functools import cached_property

class MyClass:
    def __init__(self, year: int):
        self.current_year = 2020
        self.year = year

    @cached_property  # Python >= 3.8
    def how_long_ago(self) -> int:
        return self.current_year - self.year
```

**Comment by rookwood101** [comment]
> As a workaround, you could try defining your own alias to the `property` decorator that is annotated to return `Any`. If you use your custom property decorator (that at runtime behaves exactly like `property`) for all decorated properties, mypy will silently give all such properties the `Any` type.

For others' reference, this can be done like so:
```
def property(fn) -> Any:
    """Resolves mypy 'error: Decorated property not supported'"""
    return property(fn)
```

**Comment by 18goldr** [comment]
If the system is incapable of detecting whether a decorated property works, why not just stop emitting the error, or at the very least give an option to disable this error? It's annoying to have to have a workaround for this such as adding `#type: ignore on every line that has it`. There seems to be no benefit to this error, except to say "don't trust mypy with decorated properties". Instead of emitting an error I would think it would make more sense to just note that this is a caveat to the program.

**Comment by gvanrossum** [comment]
Is there absolutely nobody reading this who would like to contribute a fix to mypy? It’s open source! Please, we can use more contributors.

**Comment by 18goldr** [comment]
> 
> 
> Is there absolutely nobody reading this who would like to contribute a fix to mypy? It’s open source! Please, we can use more contributors.

I would like to, however I don't have the time.

**Comment by Netzeband** [comment]
@18goldr You can have a look at the pull request (#8108), which was declined one year ago. It was addressing the issue that it is not possible to ignore this specific error and it introduced a regex system of ignoring error messages, which helps me a lot of making mypy silent when it comes to properties. 

Unfortunately it was declined and the maintainers are not aiming for any alternative solution here, but you can fork mypy and apply the pull-request to your fork.

**Comment by mahmoudajawad** [comment]
I'm having an issue but I'm not entirely sure it is the same as being discussed here. With the following snippet:
```python
	def archive_members(
		*, archive: tarfile.TarFile, root_path: str, search_path: str = None
	):
		l = len(f'{root_path}/')
		for member in archive.getmembers():
			if member.path.startswith(f'{root_path}/{search_path or ""}'):
				member.path = member.path[l:]
				yield member
```

mypy complains on use of `member.path` with error: `"TarInfo" has no attribute "path"`.

Now, `path` is correctly not an attribute but a decorated property method, however, I'm not getting the error mentioned here which says `Decorator not supported`, but `no attribute`. Does this fall under the same category?

**Comment by hauntsaninja** [comment]
No, it's not related. You should ask questions on Gitter. It looks like the typeshed stubs for tarfile could use some improvement (although you could instead use `name`):

<details>

```
$ stubtest tarfile --concise
tarfile.TarFile.OPEN_METH is not present in stub
tarfile.TarFile.bz2open is not present in stub
tarfile.TarFile.errors variable differs from runtime type None
tarfile.TarFile.fileobject is not present in stub
tarfile.TarFile.gzopen is not present in stub
tarfile.TarFile.taropen is not present in stub
tarfile.TarFile.xzopen is not present in stub
tarfile.TarInfo.chksum is not present in stub
tarfile.TarInfo.create_gnu_header is not present in stub
tarfile.TarInfo.create_pax_global_header is not present in stub
tarfile.TarInfo.create_pax_header is not present in stub
tarfile.TarInfo.create_ustar_header is not present in stub
tarfile.TarInfo.devmajor is not present in stub
tarfile.TarInfo.devminor is not present in stub
tarfile.TarInfo.get_info is not present in stub
tarfile.TarInfo.issparse is not present in stub
tarfile.TarInfo.linkpath is not present in stub
tarfile.TarInfo.offset is not present in stub
tarfile.TarInfo.offset_data is not present in stub
tarfile.TarInfo.path is not present in stub
tarfile.TarInfo.sparse is not present in stub
tarfile.TarInfo.tarfile is not present in stub
```
</details>

**Comment by andrewnaguib** [comment]
> Is there absolutely nobody reading this who would like to contribute a fix to mypy? It’s open source! Please, we can use more contributors.

I would, if there is a clear task description or expected behavior of what we hope to accomplish.

**Comment by JelleZijlstra** [comment]
@andrewnaguib we'd want code like that in the top post to not produce errors, and similar incorrect code (e.g., a decorator that produces a callable with too many arguments) should produce an error.

**Comment by ssbarnea** [comment]
I wondered the same few hours ago but I was lucky that I was able to just rewrite my code to avoid use of lru_cache in my case but I still have no idea how to fix it if I really need both.

**Comment by andrewnaguib** [comment]
@JelleZijlstra Would ensuring that the latest decorator (in case there are sevaral ones) returns the same type be enough?

**Comment by JelleZijlstra** [comment]
@andrewnaguib not necessarily the same type. Consider the following:

```python
T = TypeVar("T")
U = TypeVar("U")

decorator: Callable[[Callable[[T, int], U]], Callable[[T], U]]

class X:
     @property
     @decorator
     def prop(self, x: int) -> str:
           pass
```
This code is correct, because the decorated `prop` method matches what `property` wants: the decorator removes the `int` argument.

**Comment by analog-cbarber** [comment]
Seems like a simple fix would be to simply add top-level option to turn off this warning or just eliminate it entirely. What is the purpose of this warning in any case? 

It just tells us that mypy isn't smart enough to handle this code.

**Comment by ikamensh** [comment]
> Is there absolutely nobody reading this who would like to contribute a fix to mypy? It’s open source! Please, we can use more contributors.

One way to make it easier would be for someone from experienced contributors to outline high level design of how it could be done, or some areas in the code to get started. Or maybe a sketch of unit test reproducing this error in terms of internals of mypy.

**Comment by analog-cbarber** [comment]
We don't need a fancy proposal for this. Just turn off the warning or make it an actual warning rather than an error we cannot ignore without polluting our code with pragmas. What use this warning as it stands now?

Mypy should simply ignore decorators on properties for now.

**Comment by sobolevn** [comment]
I am going to try adding support for this feature. Going to send a PR in the next couple of days.

I hope it will work! 🙂

**Comment by unparalleled-js** [comment]
@sobolevn Let me know if you need help, I was confused on this warning as well.

I have code that once used  `abstractmethod` from `functools` and `mypy` never complained about that. But, when aliasing this `abstractmethod`, it suddenly does complain.

**Comment by analog-cbarber** [comment]
This is pretty much a useless error. It doesn't tell you anything about the code. It just tells you that mypy doesn't implement a feature. Why should we have to load our code up with `# type: ignore` comments for this?

Just remove this error and call it a day.

**Comment by graingert** [comment]
it looks like mypy does actually support properties on decorated functions, it just claims not to:

https://mypy-play.net/?mypy=latest&python=3.10&gist=614fc752589b60b66011de44e02d9524&flags=strict
```
from collections.abc import Callable
from typing import ParamSpec, TypeVar

P = ParamSpec("P")
T = TypeVar("T")

import functools

def some_decorator(f: Callable[P, int]) -> Callable[P, str]:
    @functools.wraps(f)
    def decorator(*args: P.args, **kwargs: P.kwargs) -> str:
        return f"{f(*args, **kwargs)}"
        
    return decorator
    
    
def some_decorator2(p: property) -> property:
    return p

class Foo:
    
    # this is a propertied decorator and fairly normal
    @property  # type: ignore[misc]
    @some_decorator
    def foo(self) -> int:
        pass
    

    # this is a decorated property and an unusual construct 
    @some_decorator2
    @property
    def bar(self) -> int:
        pass
    
    
f = Foo()
reveal_type(f.foo)
```

so there's actually two issues here. "mypy doesn't support decorated properties" and "mypy claims to not support propertied decorators when it actually does"

**Comment by leycec** [comment]
**Gah!** How is this possibly still an issue in late 2022? It's literally been eight years. Sad panda is sad.

![so sad panda](https://c.tenor.com/EWFiAD0KtbkAAAAS/sad-panda.gif)

**Comment by graingert** [comment]
> **Gah!** How is this possibly still an issue in late 2022? It's literally been eight years. Sad panda is sad.
> 
> ![so sad panda](https://camo.githubusercontent.com/0f54a59ab1ca95c60144c481e21544a604d2e6b9cdde42019d92e40be5a38579/68747470733a2f2f632e74656e6f722e636f6d2f455746694144304b74626b41414141532f7361642d70616e64612e676966) [ ![so sad panda](https://camo.githubusercontent.com/0f54a59ab1ca95c60144c481e21544a604d2e6b9cdde42019d92e40be5a38579/68747470733a2f2f632e74656e6f722e636f6d2f455746694144304b74626b41414141532f7361642d70616e64612e676966) ](https://camo.githubusercontent.com/0f54a59ab1ca95c60144c481e21544a604d2e6b9cdde42019d92e40be5a38579/68747470733a2f2f632e74656e6f722e636f6d2f455746694144304b74626b41414141532f7361642d70616e64612e676966) [ ](https://camo.githubusercontent.com/0f54a59ab1ca95c60144c481e21544a604d2e6b9cdde42019d92e40be5a38579/68747470733a2f2f632e74656e6f722e636f6d2f455746694144304b74626b41414141532f7361642d70616e64612e676966)

Afaik it's fixed in GitHub

**Comment by bew** [comment]
Any idea when will be the next release with this change?

**Comment by hauntsaninja** [comment]
Soon, follow #13385

**Comment by analog-cbarber** [comment]
So now there is just a new useless error. Now I get "Decorators on top of @property are not supported".

What purpose does this error serve?

**Comment by analog-cbarber** [comment]
That would not be correct. My decorator takes a property and returns a property. It is decorating the property and not the function.

**Comment by leycec** [comment]
**Yup.** @beartype transparently supports both modalities, too – including direct decoration of high-level property getters, setters, and deleters as well as of lower-level methods subsequently decorated as property getters, setters, and deleters. Decoration order is irrelevant (as it should be): e.g.,

```python
from beartype import beartype

class FunWithHungryBears(object):
    # This is fine.
    @beartype
    @property
    def feed_bear_garbage() -> str:
        return 'Bear responds by tearing apart garbage.'

    # This is fine, too.
    @property
    @beartype
    def throw_salmon_at_bear() -> str:
        return 'Bear responds by opening spittle-flecked muzzle.'
```

Thank Guido for `# type: ignore`. Through ignorance, we pretend mypy is sane.

**Comment by OJFord** [comment]
@bew Your suggestion was semantically different - imagine mypy didn't work with list arguments for some reason, 'try using a set instead' wouldn't really be a solution.

**Comment by samskiter** [comment]
> Soon, follow #13385

Still not fixed and appears as a gotcha in the pydantic docs: https://docs.pydantic.dev/2.0/usage/computed_fields/

(On mypy v 1.5.1)

**Comment by hhartzer** [comment]
Could someone explain why this is closed? It still seems to be an issue.
